File: Medium/Count Inversions/count-inversions.py
class Solution:
    def merge(self,nums,lo,mid,hi):
        i=lo
        j=mid+1
        temp=[]
        cnt=0
        while i<=mid and j<=hi:
            if nums[i]<=nums[j]:
                temp.append(nums[i])
                i=i+1
            else:
                temp.append(nums[j])
                cnt=cnt+(mid-i+1)
                j=j+1
        while i<=mid:
            temp.append(nums[i])
            i=i+1
        while j<=hi:
            temp.append(nums[j])
            j=j+1
        for i in range(lo,hi+1):
            nums[i]=temp[i-lo]
        self.ans=self.ans+cnt

    def mergeSort(self,nums,lo,hi):
        if lo>=hi:
            return 
        mid=(lo+hi)//2
        self.mergeSort(nums,lo,mid)
        self.mergeSort(nums,mid+1,hi)
        self.merge(nums,lo,mid,hi)
    def inversionCount(self, arr, n):
        self.ans=0
        self.mergeSort(arr,0,n-1)
        return self.ans


#{ 
 # Driver Code Starts
#Initial Template for Python 3

import atexit
import io
import sys

_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER

@atexit.register

def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())

if __name__=='__main__':
    t = int(input())
    for tt in range(t):
        n = int(input())
        a = list(map(int, input().strip().split()))
        obj = Solution()
        print(obj.inversionCount(a,n))
# } Driver Code Ends

File: Difficulty: Medium/Count Inversions/count-inversions.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.lang.*;
import java.util.*;

class Sorting {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        long t = Integer.parseInt(br.readLine());
        for (int g = 0; g < t; g++) {
            String[] str = (br.readLine()).trim().split(" ");
            long arr[] = new long[str.length];
            for (int i = 0; i < str.length; i++) arr[i] = Integer.parseInt(str[i]);
            System.out.println(new Solution().inversionCount(arr));
        }
    }
}

// } Driver Code Ends


// User function Template for Java

class Solution {
    
    
    static int merge(long []arr,int low,int mid,int high){
        int i=low;
        int j=mid+1;
        ArrayList<Long> temp=new ArrayList<>();
        int cnt=0;
        while(i<=mid && j<=high){
            if(arr[i]<=arr[j]){
                temp.add(arr[i]);
                i=i+1;
            }else{
                temp.add(arr[j]);
                cnt=cnt+(mid-i+1);
                j=j+1;
            }
        }
        
        while(i<=mid){
            temp.add(arr[i]);
            i=i+1;
        }
        while(j<=high){
            temp.add(arr[j]);
            j=j+1;
        }
        
        for(int k=low;k<=high;k++){
            arr[k]=temp.get(k-low);
        }
        return cnt;
    }
    
    static int mergeSort(long []arr,int low, int high){
        int cnt=0;
        if(low>=high){
            return cnt;
        }
        int mid=(low+high)/2;
        cnt+=mergeSort(arr,low,mid);
        cnt+=mergeSort(arr,mid+1,high);
        cnt+=merge(arr,low,mid,high);
        return cnt;
    }
    static long inversionCount(long arr[]) {
        int low=0;
        int high=arr.length-1;
        return mergeSort(arr,0,high);
        
    }
}